### ISSUE TITLE
ImportError on PEP 420 namespace

### ISSUE THREAD

**Issue by JulienPalard** [PyCQA/pylint#1515]
I'm experiencing a strange bug probably in relation of PEP 420 - Implicit Namespace Packages.

Reproduction:

I'm on the actual tip, git status clean:
```bash
julien@dev2 (master)$ git lg | head -n 1
* 6f2b4cc - (HEAD, origin/master, origin/HEAD, master) Add a warning for bad imports from the types module (#1501) (3 days ago) <Roy Williams>
julien@dev2 (master)$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```

I enter a new, clean venv, in which I install pylint and protobuf (itself using a PEP 420 implicit namespace package "google"):
```bash
julien@dev2 (master)$ workon tmp
(tmp) julien@dev2 (master)$ python3 -m pip install . 
Processing /tmp/pylint
Collecting astroid>=1.5.1 (from pylint==1.8.0)
  Using cached astroid-1.5.2-py2.py3-none-any.whl
Collecting six (from pylint==1.8.0)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting isort>=4.2.5 (from pylint==1.8.0)
  Using cached isort-4.2.9-py2.py3-none-any.whl
Collecting mccabe (from pylint==1.8.0)
  Using cached mccabe-0.6.1-py2.py3-none-any.whl
Collecting lazy-object-proxy (from astroid>=1.5.1->pylint==1.8.0)
  Using cached lazy_object_proxy-1.3.1-cp35-cp35m-manylinux1_x86_64.whl
Collecting wrapt (from astroid>=1.5.1->pylint==1.8.0)
Installing collected packages: lazy-object-proxy, six, wrapt, astroid, isort, mccabe, pylint
  Found existing installation: pylint 2.0.0
    Not uninstalling pylint at /tmp/pylint, outside environment /home/julien/.venvs/tmp
  Running setup.py install for pylint ... done
Successfully installed astroid-1.5.2 isort-4.2.9 lazy-object-proxy-1.3.1 mccabe-0.6.1 pylint-1.8.0 six-1.10.0 wrapt-1.10.10
(tmp) julien@dev2 (master)$ python3 -m pip install protobuf
Collecting protobuf
  Using cached protobuf-3.3.0-cp35-cp35m-manylinux1_x86_64.whl
Requirement already satisfied: setuptools in /home/julien/.venvs/tmp/lib/python3.5/site-packages (from protobuf)
Requirement already satisfied: six>=1.9 in /home/julien/.venvs/tmp/lib/python3.5/site-packages (from protobuf)
Installing collected packages: protobuf
Successfully installed protobuf-3.3.0
```

I create a test directory which a test file:
```bash
(tmp) julien@dev2 (master)$ cd /tmp/
(tmp) julien@dev2$ mkdir foobar
(tmp) julien@dev2$ cd foobar
(tmp) julien@dev2$ cat > foobar.py
from google.protobuf import descriptor as _descriptor

_descriptor.FileDescriptor()
```

I ensure pylint hits pylint from the venv:
```bash
(tmp) julien@dev2$ hash -r
```

and run it:
```bash
(tmp) julien@dev2$ pylint foobar.py
No config file found, using default configuration
************* Module foobar
C:  1, 0: Missing module docstring (missing-docstring)
Traceback (most recent call last):
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/manager.py", line 129, in ast_from_module_name
    module = modutils.load_module_from_name(modname)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/modutils.py", line 190, in load_module_from_name
    return load_module_from_modpath(dotted_name.split('.'), path, use_sys)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/modutils.py", line 232, in load_module_from_modpath
    mp_file, mp_filename, mp_desc = imp.find_module(part, path)
  File "/usr/lib/python3.5/imp.py", line 296, in find_module
    raise ImportError(_ERR_MSG.format(name), name=name)
ImportError: No module named 'google'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/julien/.venvs/tmp/bin/pylint", line 11, in <module>
    load_entry_point('pylint==1.8.0', 'console_scripts', 'pylint')()
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/__init__.py", line 13, in run_pylint
    Run(sys.argv[1:])
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/lint.py", line 1308, in __init__
    linter.check(args)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/lint.py", line 734, in check
    self._do_check(files_or_modules)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/lint.py", line 863, in _do_check
    self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/lint.py", line 942, in check_astroid_module
    walker.walk(ast_node)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/utils.py", line 980, in walk
    self.walk(child)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/utils.py", line 980, in walk
    self.walk(child)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/utils.py", line 977, in walk
    cb(astroid)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/checkers/base.py", line 556, in visit_call
    abstract_methods = _has_abstract_methods(infered)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/checkers/base.py", line 209, in _has_abstract_methods
    return len(utils.unimplemented_abstract_methods(node)) > 0
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/checkers/utils.py", line 552, in unimplemented_abstract_methods
    infered = safe_infer(obj)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/pylint/checkers/utils.py", line 773, in safe_infer
    value = next(inferit)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/bases.py", line 95, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/context.py", line 40, in cache_generator
    for result in generator:
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/decorators.py", line 138, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/inference.py", line 237, in infer_attribute
    for owner in self.expr.infer(context):
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/context.py", line 40, in cache_generator
    for result in generator:
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/bases.py", line 95, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/context.py", line 40, in cache_generator
    for result in generator:
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/decorators.py", line 82, in wrapped
    generator = _func(node, context, **kwargs)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/inference.py", line 226, in infer_import_from
    stmts = module.getattr(name, ignore_locals=module is self.root())
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/scoped_nodes.py", line 345, in getattr
    result = [self.import_module(name, relative_only=True)]
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/scoped_nodes.py", line 409, in import_module
    return MANAGER.ast_from_module_name(absmodname)
  File "/home/julien/.venvs/tmp/lib/python3.5/site-packages/astroid/manager.py", line 133, in ast_from_module_name
    modname=modname, path=spec.location, error=ex))
AttributeError: module 'astroid.interpreter._import.spec' has no attribute 'location'
```

Now, thanks to git bisect, I found some strange things: chronologically, from the oldest to the newest, sha1 ac33f42 has the bug, d1ca5d8 don't have it, then 28e04bd have it back.

Strange, is that 28e04bd, reintroducing the bug, is only a change in version, if I "revert" it:
```bash
(tmp) julien@dev2$ sed -i 's/(1, 8, 0)/(2, 0, 0)/' ../pylint/pylint/__pkginfo__.py 
```

And try again with a *clean* venv:
```bash
(tmp) julien@dev2$ deactivate 
julien@dev2$ rm -fr ~/.venvs/tmp/
julien@dev2$ workon tmp/
(tmp) julien@dev2$ cd ../pylint/
(tmp) julien@dev2 (master)$ python3 -m pip install . 
Processing /tmp/pylint
  Requirement already satisfied (use --upgrade to upgrade): pylint==2.0.0 from file:///tmp/pylint in /tmp/pylint
Collecting astroid>=1.5.1 (from pylint==2.0.0)
  Using cached astroid-1.5.2-py2.py3-none-any.whl
Collecting six (from pylint==2.0.0)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting isort>=4.2.5 (from pylint==2.0.0)
  Using cached isort-4.2.9-py2.py3-none-any.whl
Collecting mccabe (from pylint==2.0.0)
  Using cached mccabe-0.6.1-py2.py3-none-any.whl
Collecting lazy-object-proxy (from astroid>=1.5.1->pylint==2.0.0)
  Using cached lazy_object_proxy-1.3.1-cp35-cp35m-manylinux1_x86_64.whl
Collecting wrapt (from astroid>=1.5.1->pylint==2.0.0)
Installing collected packages: lazy-object-proxy, wrapt, six, astroid, isort, mccabe
Successfully installed astroid-1.5.2 isort-4.2.9 lazy-object-proxy-1.3.1 mccabe-0.6.1 six-1.10.0 wrapt-1.10.10
(tmp) julien@dev2 (master)$ python3 -m pip install protobuf
Collecting protobuf
  Using cached protobuf-3.3.0-cp35-cp35m-manylinux1_x86_64.whl
Requirement already satisfied: setuptools in /home/julien/.venvs/tmp/lib/python3.5/site-packages (from protobuf)
Requirement already satisfied: six>=1.9 in /home/julien/.venvs/tmp/lib/python3.5/site-packages (from protobuf)
Installing collected packages: protobuf
Successfully installed protobuf-3.3.0
```

And try again, it works:
```
(tmp) julien@dev2 (master)$ cd -
/tmp/foobar
(tmp) julien@dev2$ pylint foobar.py
No config file found, using default configuration
************* Module foobar
C:  1, 0: Missing module docstring (missing-docstring)
E:  1, 0: Unable to import 'google.protobuf' (import-error)


Report
======
2 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |1      |1          |=          |0.00        |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |0      |0          |=          |0           |0        |
+---------+-------+-----------+-----------+------------+---------+
|method   |0      |0          |=          |0           |0        |
+---------+-------+-----------+-----------+------------+---------+
|function |0      |0          |=          |0           |0        |
+---------+-------+-----------+-----------+------------+---------+



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |3      |60.00 |NC       |NC         |
+----------+-------+------+---------+-----------+
|docstring |0      |0.00  |NC       |NC         |
+----------+-------+------+---------+-----------+
|comment   |0      |0.00  |NC       |NC         |
+----------+-------+------+---------+-----------+
|empty     |2      |40.00 |NC       |NC         |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |1      |2        |-1.00      |
+-----------+-------+---------+-----------+
|refactor   |0      |0        |=          |
+-----------+-------+---------+-----------+
|warning    |0      |0        |=          |
+-----------+-------+---------+-----------+
|error      |1      |1        |=          |
+-----------+-------+---------+-----------+



Messages
--------

+------------------+------------+
|message id        |occurrences |
+==================+============+
|missing-docstring |1           |
+------------------+------------+
|import-error      |1           |
+------------------+------------+



Global evaluation
-----------------
Your code has been rated at -20.00/10 (previous run: -25.00/10, +5.00)
```

Clearly, at one point, "imp.find_module" is called like:
```
>>> import imp
>>> imp.find_module('google')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/julien/.venvs/tmp/lib/python3.5/imp.py", line 296, in find_module
    raise ImportError(_ERR_MSG.format(name), name=name)
ImportError: No module named 'google'
>>> import google  # Just to proove I had it at this moment
>>>
```

But it may seems legitimate that find_modules don't handle PEP 420 packages, the tricky question is why changing the pylint version number changes this, I tried to grep for "if version …" but found nothing interesting (yet it's something like that, someone here may now more).

Last details: I use Python 3.5.3 and my "workon" is not from virtualenvwrapper but a minimalist
```
$ type workon
workon is a function
workon () 
{ 
    local VENVS="$HOME/.venvs";
    if [ -z "$1" -o z"$1" = z"." ]; then
        local VENV_NAME="$(basename "$PWD")";
    else
        local VENV_NAME="$1";
    fi;
    [ -d "$VENVS/$VENV_NAME" ] || mkdir -p "$VENVS/$VENV_NAME";
    [ -f "$VENVS/$VENV_NAME/bin/activate" ] || python3 -m venv "$VENVS/$VENV_NAME";
    . "$VENVS/$VENV_NAME/bin/activate"
}
```
It it helps to reproduce my reproduction history.

**Comment by evanj** [comment]
I have what might be the same issue, also with the `google.protobuf` package. Although in my case, the errors are about not finding *subpackages* of `google.protobuf`, and it complains about them relative to where they are imported. I'm using Python 2.7.9 on Debian. Here is the script to reproduce the problem. It creates a new virtualenv, installs the latest versions of pylint and protobuf, runs a script that shows that the imports work, then runs Pylint which fails to find the imports:

```
virtualenv bugvenv
bugvenv/bin/pip install pylint protobuf
cat > hello.py <<HERE
import google.protobuf.reflection
print 'google.protobuf.reflection:', google.protobuf.reflection
HERE
bugvenv/bin/python hello.py
bugvenv/bin/pylint --errors-only hello.py
```

Output of the script:

```
$ bugvenv/bin/python hello.py 
google.protobuf.reflection: <module 'google.protobuf.reflection' from '/home/evan_jones/bugvenv/local/lib/python2.7/site-packages/google/protobuf/reflection.pyc'>
```

Output from Pylint:

```
$ bugvenv/bin/pylint --errors-only hello.py
No config file found, using default configuration
************* Module hello
E:  1, 0: No name 'reflection' in module 'hello.google.protobuf' (no-name-in-module)
E:  2,37: Module 'hello.google.protobuf' has no 'reflection' member (no-member)
```

To work around this, I can add `from __future__ import absolute_imports` to the script. In my actual program, that fixes one of the errors, but then I still get errors about `No name 'cloud' in module 'google' (no-name-in-module)` from `import google.cloud.storage`, so something weird is still happening.

**Comment by evanj** [comment]
This may be a duplicate of #1315

**Comment by PCManticore** [comment]
@evanj and @JulienPalard I can't reproduce both issues with pylint 1.8.0 and latest astroid. Can you try again with it and see if it reproduces for you or not?

**Comment by evanj** [comment]
I am still able to reproduce it, although now it appears it needs to be *nested* virtualenvs, and only happens for python2.

# Steps to reproduce

1. Clone latest astroid (cf5e65dcb1) and pylint (0b1b25a63f)
2. Create a python2 virtualenv called `bugvenv2` (I used Python 2.7.12 from Ubuntu 16.04)
3. Run setup.py to install astroid and pylint into the virtualenvs
4. Run the following:

```
bugvenv2/bin/pip install google-cloud-core
cat > importbug.py <<HERE
import google.cloud.exceptions
print('google.cloud.exceptions:', google.cloud.exceptions)
HERE
bugvenv2/bin/python2 importbug.py
bugvenv2/bin/pylint importbug.py
```

Output from pylint:

```
E:  1, 0: No name 'cloud' in module 'google' (no-name-in-module)
E:  1, 0: Unable to import 'google.cloud.exceptions' (import-error)
I:  2,34: Module 'google' has not 'cloud' member, but source is unavailable. Consider adding this module to extension-package-whitelist if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
```

**Comment by PCManticore** [comment]
But if `google.cloud` is a namespace package as specified by PEP 420, how exactly would it work on Python 2, since that feature wasn't backported?

**Comment by evanj** [comment]
I am not an expert here. My understanding is that in `setup.py` when you declare `namespace_packages`, `setuptools` creates a `.pth` file that gets shipped with the package, into the `site-packages` folder. This causes Python to automatically import that `.pth` at startup. In the specific case of `google-cloud-core`, they declare the following:

https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/core/setup.py#L71

```
    namespace_packages=[
        'google',
        'google.cloud',
        'google.api',
    ],
```

This causes the file `site-packages/google_cloud_core-0.27.1-py3.6-nspkg.pth` to get created, which contains generated code that looks like:

```
import sys, types, os;pep420 = sys.version_info > (3, 3);p = os.path.join(sys._getframe(1).f_locals['sitedir'], *('google', 'cloud'));ie = os.path.exists(os.path.join(p,'__init__.py'));m = not ie and not pep420 and sys.modules.setdefault('google.cloud', types.ModuleType('google.cloud'));mp = (m or []) and m.__dict__.setdefault('__path__',[]);(p not in mp) and mp.append(p);m and setattr(sys.modules['google'], 'cloud', m)
```

(there are 5 lines like that in this file, one for each "namespace" part in the `namespace_packages` declaration from setup.py)

**Comment by PCManticore** [comment]
I see. In this case, this needs to be in a separate issue, since it's not an implicit namespace package as mandated by PEP 420 (and a potential fix might be completely different)

**Comment by evanj** [comment]
Would you like me to create the issue? I am happy to do that!

**Comment by PCManticore** [comment]
Sure thing! Please do! :)

**Comment by evanj** [comment]
Created https://github.com/PyCQA/pylint/issues/1686 for this new issue.
